qualtran/bloqs/arithmetic/comparison.py

```diff
@@ -947,6 +948,96 @@ def _gt_k() -> GreaterThanConstant:
 _GREATER_THAN_K_DOC = BloqDocSpec(bloq_cls=GreaterThanConstant, examples=[_gt_k])
 
 
+@frozen
+class Equals(Bloq):
+    r"""Implements |x>|y>|t> => |x>|y>|t ⨁ (x = y)> using $n-1$ Toffoli gates.
+    Args:
+        dtype: Data type of the input registers `x` and `y`.
+    Registers:
+        x: Bitsize-sized input register.
+        y: Bitsize-sized input register.
+        target: Register to hold result of comparison.
+    """
+
+    dtype: QDType
+
+    @cached_property
+    def signature(self) -> Signature:
+        return Signature.build_from_dtypes(x=self.dtype, y=self.dtype, target=QBit())
+
+    @cached_property
+    def bitsize(self) -> SymbolicInt:
+        return self.dtype.num_qubits
+
+    def is_symbolic(self):
+        return is_symbolic(self.dtype)
+
+    def wire_symbol(self, reg: Optional[Register], idx: Tuple[int, ...] = tuple()) -> WireSymbol:
+        if reg is None:
+            return Text("")
+        if reg.name == 'x':
+            return TextBox("In(x)")
+        if reg.name == 'y':
+            return TextBox("In(y)")
+        elif reg.name == 'target':
+            return TextBox("⨁(x = y)")
+        raise ValueError(f'Unknown register symbol {reg.name}')
+
+    def build_composite_bloq(
+        self, bb: 'BloqBuilder', x: 'Soquet', y: 'Soquet', target: 'Soquet'
+    ) -> Dict[str, 'SoquetT']:
+        if self.is_symbolic():
+            raise DecomposeTypeError(f"cannot decompose symbolic {self}")
+
+        x, y = bb.add(Xor(self.dtype), x=x, y=y)
+        if self.bitsize == 1:
+            y = bb.add(XGate(), q=y)
+            y, target = bb.add(CNOT(), ctrl=y, target=target)
+            y = bb.add(XGate(), q=y)
+        elif self.bitsize == 2:
+            y_split = bb.split(y)
+            y_split, out = bb.add(And(0, 0), ctrl=y_split)
+            out, target = bb.add(CNOT(), ctrl=out, target=target)
+            y_split = bb.add(And(0, 0).adjoint(), ctrl=y_split, target=out)
+            y = bb.join(y_split, self.dtype)
+        else:
+            y_split = bb.split(y)
+            y_split, junk, out = bb.add(MultiAnd(cvs=[0] * self.bitsize), ctrl=y_split)
+            out, target = bb.add(CNOT(), ctrl=out, target=target)
+            y_split = bb.add(
+                MultiAnd(cvs=[0] * self.bitsize).adjoint(), ctrl=y_split, junk=junk, target=out
+            )
+            y = bb.join(y_split, self.dtype)
```

Comment:
You can use the `MultiControlX` bloq to simplify the decomposition: https://github.com/quantumlib/Qualtran/blob/db8b9938d8210024bb6b92942376363586165efc/qualtran/bloqs/mcmt/multi_control_pauli.py#L183